virtool/indexes/data.py

```diff
@@ -0,0 +1,303 @@
+            user = await self._db.users.find_one(agg["user"]["id"])
+            ready_indexes.append(
+                {
+                    "id": agg["index"],
+                    "version": agg["version"],
+                    "reference": {
+                        "id": agg["_id"],
+                    },
+                    "change_count": agg["change_count"],
+                    "created_at": agg["created_at"],
+                    "has_files": agg["has_files"],
+                    "modified_otu_count": agg["modified_otu_count"],
+                    "user": {
+                        "id": user["_id"],
+                        "handle": user["handle"],
+                        "administrator": user["administrator"],
+                    },
+                    "ready": agg["ready"],
+                    "job": agg["job"],
+                }
+            )
+
+        return [IndexMinimal(**index) for index in ready_indexes]
+
+    async def get(self, index_id: str) -> Index:
+        """
+        Get a single index by its ID.
+
+        :param index_id: the index ID
+        :return: the index
+        """
+        document = await self._db.indexes.find_one(index_id)
+
+        if not document:
+            raise ResourceNotFoundError()
+
+        contributors, otus = await asyncio.gather(
+            virtool.history.db.get_contributors(self._db, {"index.id": index_id}),
+            virtool.indexes.db.get_otus(self._db, index_id),
+        )
+
+        document.update(
+            {
+                "change_count": sum(v["change_count"] for v in otus),
+                "contributors": contributors,
+                "otus": otus,
+            }
+        )
+
+        document = await virtool.indexes.db.attach_files(
+            self._pg, self._config.base_url, document
+        )
+
+        document = await virtool.indexes.db.processor(self._db, document)
+
+        return Index(**document)
+
+    async def get_reference(self, index_id: str) -> ReferenceNested:
+        """
+        Get a reference associated with an index.
+
+        :param index_id: the index ID
+        :return: the reference
+        """
+        index = await self._db.indexes.find_one(index_id)
+
+        if index is None:
+            raise ResourceNotFoundError()
+
+        reference = index["reference"]
+
+        return ReferenceNested(**reference)
+
+    async def get_json_path(self, index_id: str) -> Path:
+        """
+        Get the json path needed for a complete compressed JSON
+        representation of the index OTUs.
+
+        :param index_id: the index ID
+        :return: the json path
+        """
+        index = await self._db.indexes.find_one(index_id)
+
+        if index is None:
+            raise ResourceNotFoundError()
+
+        ref_id = index["reference"]["id"]
+
+        json_path = (
+            join_index_path(self._config.data_path, ref_id, index_id) / "otus.json.gz"
+        )
+
+        if not json_path.exists():
```

Comment:
If the filesystem is under load this might block the application unless we use `run_in_thread`.
```suggestion
        if not await run_in_thread(json_path.exists):
```